micrometer-commons/src/main/java/io/micrometer/common/util/StreamUtils.java

```diff
@@ -0,0 +1,31 @@
+/*
+ * Copyright 2024 VMware, Inc.
+ *
+ * Licensed under the Apache License, Version 2.0 (the "License");
+ * you may not use this file except in compliance with the License.
+ * You may obtain a copy of the License at
+ *
+ * https://www.apache.org/licenses/LICENSE-2.0
+ *
+ * Unless required by applicable law or agreed to in writing, software
+ * distributed under the License is distributed on an "AS IS" BASIS,
+ * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+ * See the License for the specific language governing permissions and
+ * limitations under the License.
+ */
+package io.micrometer.common.util;
+
+import java.util.Map;
+import java.util.concurrent.ConcurrentHashMap;
+import java.util.function.Function;
+import java.util.function.Predicate;
+
+public class StreamUtils {
+
+    public static <T> Predicate<T> distinctByKey(Function<? super T, ?> keyExtractor) {
```

Comment:
If this is only used in the one place, I'd rather not add it as public API. We can keep it private in AnnotationHandler, I think.

Reply:
honestly i would not agree. It's purely independent code which can and should be extracted into own utility class. I would even imagine that it can be later widely reused elsewhere.
Regarding leaving it in AnnotationHandler - technically it would also work fine, however that one has quite enough code dealing with annotations and  reflection, so messing it with stream API utility methods would make purpose of the class even less obvious.
In any case if you think that it is necessary then i will move the method to AnnotationHandler. Just let me know it's a final "go|no go" unblocker :)

Reply:
I think your approach would make sense in application code that is not shared like this library with many applications. We need to support public API we expose and we do not want to pollute users' classpaths with utility classes they don't expect from an observability library (we already unfortunately do this with yet another StringUtils on everyone's classpath) for something we only want to use internally.

> I would even imagine that it can be later widely reused elsewhere.

We have not had a need for it elsewhere so far, and if we eventually do have such a need, we can revisit the decision. It's easy to move private code and make it public, but it is not easy to remove or change public API that may be used by many people.

I'll take care of moving this with merging so we can get this enhancement in the soon to be released milestone.

Reply:
oh, i haven't looked at it from this perspective, thanks for the detailed explanation. `utility classes they don't expect from an observability library` - that is more than super-descriptive statement.